src/jit/codegenxarch.cpp

```diff
@@ -7049,6 +7050,12 @@ void CodeGen::genCompareLong(GenTreePtr treeNode)
     // Emit the compare instruction
     getEmitter()->emitInsBinary(ins, cmpAttr, hiOp1, hiOp2);
 
+    // If the result is not being materialized in a register, we're done.
+    if (targetReg == REG_NA)
+    {
+        return;
+    }
+
```

Comment:
> Even in case (2), GT_CMP will not marked as contained and it will need to materialize the result into a register.

This seems like an oversight: why should the result be materialized into a register if it is not used? All that we need to do in the case that the result is unused is execute any side-effects associated with the compare (in particular, we need to dereference any memory operands). This is also why the code bails out after the high compare: I would be surprised if we ever load the high and low parts of a GT_LONG from different base pointers (i.e. if loading the high half causes an AV, so should loading the low half and vice-versa), so I believe that it is sufficient to only execute the high part of the compare. It may be worth dereferencing the low half as well for the sake of paranoia (or  at a minimum better validating the previously mentioned assumption).